src/timer.tsx

```diff
@@ -0,0 +1,438 @@
+    const [currentSeconds, setCurrentSeconds] = React.useState(remainingTime.seconds);
+
+    useIsomorphicLayoutEffect(() => {
+        let intervalId: NodeJS.Timeout;
+
+        const updateCurrentTime = () => {
+            const remainingTime = getRemainingTime(endTimestamp);
+            setCurrentDays(remainingTime.days);
+            setCurrentHours(remainingTime.hours);
+            setCurrentMinutes(remainingTime.minutes);
+            setCurrentSeconds(remainingTime.seconds);
+
+            // Stop computing values if there is no time remaining
+            if (
+                !remainingTime.days &&
+                !remainingTime.hours &&
+                !remainingTime.minutes &&
+                !remainingTime.seconds
+            ) {
+                clearInterval(intervalId);
+            }
+        };
+
+        if (!isRunningAcceptanceTest()) {
+            updateCurrentTime();
+            intervalId = setInterval(updateCurrentTime, SECOND_IN_MS);
+            return () => clearInterval(intervalId);
+        }
+    }, [endTimestamp]);
+
+    const shouldRenderDays = shouldRenderUnit('days', labelType, minTimeUnit, maxTimeUnit);
+    const shouldRenderHours = shouldRenderUnit('hours', labelType, minTimeUnit, maxTimeUnit);
+    const shouldRenderMinutes = shouldRenderUnit('minutes', labelType, minTimeUnit, maxTimeUnit);
+
+    const maximumRenderedUnit = shouldRenderDays
+        ? 'days'
+        : shouldRenderHours
+        ? 'hours'
+        : shouldRenderMinutes
+        ? 'minutes'
+        : 'seconds';
+
+    const days = currentDays;
+
+    // if hours is the maximum unit, add remaining days
+    const hours = maximumRenderedUnit === 'hours' ? currentHours + days * DAY_IN_HOURS : currentHours;
+
+    // if minutes is the maximum unit, add remaining days and hours
+    const minutes =
+        maximumRenderedUnit === 'minutes'
+            ? currentMinutes + HOUR_IN_MINUTES * (days * DAY_IN_HOURS + hours)
+            : currentMinutes;
+
+    // if minutes is the maximum unit, add remaining days, hours and minutes
+    const seconds =
+        maximumRenderedUnit === 'seconds'
+            ? currentSeconds +
+              MINUTE_IN_SECONDS * (days * DAY_IN_HOURS * HOUR_IN_MINUTES + hours * HOUR_IN_MINUTES + minutes)
+            : currentSeconds;
+
+    const [timerValue, setTimerValue] = React.useState(
+        getFilteredTimerValue({days, hours, minutes, seconds}, labelType, minTimeUnit, maxTimeUnit)
+    );
+
+    React.useEffect(() => {
+        const currentTimerValue = getFilteredTimerValue(
+            {days, hours, minutes, seconds},
+            labelType,
+            minTimeUnit,
+            maxTimeUnit
+        );
+
+        if (!isEqual(currentTimerValue, timerValue)) {
+            setTimerValue(currentTimerValue);
+            const timestampValue: Timestamp = {};
+            currentTimerValue.forEach((item) => (timestampValue[item.unit] = item.value));
+            onProgress?.(timestampValue);
```

Comment:
why don't we call onProgress when we set the remaining time state?

Reply:
Discussed offline. We call setState every second to update timer value internally and recompute everything. But if minTimeUnit is "minutes", we don't want to call onProgress every second